include/exadg/structure/preconditioners/multigrid_preconditioner.cpp

```diff
@@ -126,6 +126,35 @@ MultigridPreconditioner<dim, Number>::update()
   }
 }
 
+template<int dim, typename Number>
+void
+MultigridPreconditioner<dim, Number>::initialize_dof_handler_and_constraints(
+  bool const                         operator_is_singular,
+  dealii::FiniteElement<dim> const & fe,
+  Map_DBC const &                    dirichlet_bc,
+  Map_DBC_ComponentMask const &      dirichlet_bc_component_mask)
+{
+  Base::initialize_dof_handler_and_constraints(operator_is_singular,
+                                               fe,
+                                               dirichlet_bc,
+                                               dirichlet_bc_component_mask);
+
+  // additional constraints without Dirichlet degrees of freedom:
+  // Due to the interface of the base class MultigridPreconditionerBase, we also need to set up
+  // additional DoFHandler objects.
+  if(nonlinear)
+  {
+    Map_DBC               dirichlet_bc_empty;
+    Map_DBC_ComponentMask dirichlet_bc_empty_component_mask;
+    this->do_initialize_dof_handler_and_constraints(false,
+                                                    fe,
+                                                    dirichlet_bc_empty,
+                                                    dirichlet_bc_empty_component_mask,
+                                                    dof_handlers_inhomogeneous,
+                                                    constraints_inhomogeneous);
+  }
```

Comment:
Just for clarity: We need this additional step because we might interpolate some fields to the multigrid levels, where using the Dirichlet boundary conditions are wrong. In case we had a different/better way to transfer the linearization point (or, if it ever gets relevant, a possible ALE mesh deformation we might impose in a structural problem), we would not need these additions, right? That is certainly a different problem for another pull request, but I think it is at least very clear where these additional data structures would need to be eliminated if we do not need them any more.

Reply:
I do not fully understand this comment. Maybe let's discuss this in person.

Reply:
Please also note commit https://github.com/exadg/exadg/pull/506/commits/c3dc80b3fed2e58b759358138869363fa0d55733